packages/mds-collector-backend/service/provider.ts

```diff
@@ -0,0 +1,151 @@
+    return { $schema: 'http://json-schema.org/draft-07/schema#', ...schema }
+  } catch (error) {
+    throw typeof error === 'object' && error !== null && error.code === 'MODULE_NOT_FOUND'
+      ? new NotFoundError(`Schema module ${module} not found`)
+      : error
+  }
+}
+
+const getSchemaObject = async (schema_id: string): Promise<SchemaObject> => {
+  const $schema = (schema: SchemaObject) => ({
+    $schema: 'http://json-schema.org/draft-07/schema#',
+    ...schema
+  })
+
+  const schema = SchemaObjects.get(schema_id) ?? $schema(await importSchemaObject(schema_id))
+  if (!SchemaObjects.has(schema_id)) {
+    SchemaObjects.set(schema_id, schema)
+  }
+  return schema
+}
+
+const getValidateFunction = async (schema_id: string): Promise<CollectorValidateFunction> => {
+  const validator = ValidateFunctions.get(schema_id) ?? SchemaValidator(await getSchemaObject(schema_id))
+  if (!ValidateFunctions.has(schema_id)) {
+    ValidateFunctions.set(schema_id, validator)
+  }
+  return validator
+}
+
+const MockStreamProducer: CollectorStreamProducer = {
+  initialize: async () => undefined,
+  shutdown: async () => undefined,
+  write: async message => undefined
+}
+
+const createStreamProducer = async (schema_id: string): Promise<CollectorStreamProducer> => {
+  const topic = `${TENANT_ID}.collector.${schema_id}`
+  // TODO: Do we need to create the topic?
+  const producer = process.env.KAFKA_HOST !== undefined ? stream.KafkaStreamProducer(topic) : MockStreamProducer
+  await producer.initialize()
+  return producer
+}
+
+const getStreamProducer = async (schema_id: string): Promise<CollectorStreamProducer> => {
+  const producer = StreamProducers.get(schema_id) ?? (await createStreamProducer(schema_id))
+  if (!StreamProducers.has(schema_id)) {
+    StreamProducers.set(schema_id, producer)
+  }
+  return producer
+}
+
+export const CollectorServiceProvider: ServiceProvider<CollectorService> & ProcessController = {
+  start: CollectorRepository.initialize,
+  stop: CollectorRepository.shutdown,
+  getMessageSchema: async schema_id => {
+    try {
+      const schema = await getSchemaObject(schema_id)
+      return ServiceResult(schema)
+    } catch (error) {
+      const exception = ServiceException(`Error Reading Schema ${schema_id}`, error)
+      logger.error(exception, error)
+      return exception
+    }
+  },
+  writeSchemaMessages: async (schema_id, producer_id, messages) => {
+    try {
+      const [validate, producer] = await Promise.all([getValidateFunction(schema_id), getStreamProducer(schema_id)])
+
+      const invalid = messages.reduce<{ position: number; errors: Partial<ErrorObject>[] }[]>(
+        (errors, message, position) => {
+          // eslint-reason validate function has previously verified that errors is non-null
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          return validate(message) ? errors : errors.concat({ position, errors: validate.errors! })
```

Comment:
I recommend we use the AJV `async` validators here. The synchronous API is honestly pretty abysmal (as demonstrated by the need to do a non-null assertion here)

Reply:
The `async` API is kinda weird too. I'll look into abstracting the Ajv API in the SchemaValidator to provide a better interface. We can avoid the non-null assertion without resorting to async validation.